pkg/config/model/types.go

```diff
@@ -111,7 +111,15 @@ type Setup interface {
 	SetEnvPrefix(in string)
 	BindEnv(input ...string)
 	SetEnvKeyReplacer(r *strings.Replacer)
+
+	// SetEnvKeyTransformer is deprecated in favor of LoadEnv* functions
 	SetEnvKeyTransformer(key string, fn func(string) interface{})
+	// The following helpers allow a type to be enforce when parsing environment variables. Most of them exists to
```

Comment:
Overall, I think this PR is really good for moving the API towards requiring specific types to be declared. I'm still a bit bothered that it allows for arbitrary parsing functions to be used. Though we can't avoid that due to how Viper works (it waits until later to parse the env vars, and thus has to use callbacks), I wonder if it would be an improvement to move the parsing functions into one centralized location to at least trim-down the API's surface area and help to consolidate knowledge of how the config works into less places.

I'm imagining a package at `comp/core/config/configsyntax` that would contain the functions like `parseKVList` and `splitCSVString`. Maybe that's too big of a change for this PR, which is just trying to improve the type-safety. If you think it's worth doing this at some point, maybe just add a TODO for now to note that this code isn't final.

Reply:
I agree and it was the first thing I tried. But I quickly realized that there is a lot of very custom logic in those functions. For example we have 4 different format+logic just to parse a list of strings. Which made me think that what we really want to enforce is the format returned by the function.

One thing we could still do in another PR is isolate the more generic function into a helper folder to promote their use in the future.

Reply:
Yeah I think just having the parser functions as helpers in their own package is good enough. And it's fine doing that in a follow-up PR.